shell/platform/darwin/macos/framework/Source/FlutterEngine.mm

```diff
@@ -408,39 +464,77 @@ - (void)loadAOTData:(NSString*)assetsDir {
+  NSAssert(controller != nil, @"The controller must not be nil.");
+  NSAssert(![controller attached],
+           @"The incoming view controller is already attached to an engine.");
+  NSAssert([_viewControllers objectForKey:@(viewId)] == nil, @"The requested view ID is occupied.");
+  // The first view controller is guaranteed to have the fixed default view ID.
+  NSAssert([_viewControllers count] != 0 || viewId == kFlutterDefaultViewId,
+           @"The first view controller should have the fixed default view ID, but is "
+           @"given %lld.",
+           viewId);
+  [controller attachToEngine:self withId:viewId];
+  NSAssert(controller.id == viewId, @"Failed to assign view ID.");
+  [_viewControllers setObject:controller forKey:@(viewId)];
+}
+
+- (void)deregisterViewControllerForId:(uint64_t)viewId {
+  FlutterViewController* oldController = [self viewControllerForId:viewId];
+  if (oldController != nil) {
+    [oldController detachFromEngine];
+    [_viewControllers removeObjectForKey:@(viewId)];
+  }
+}
+
+- (void)shutDownIfNeeded {
+  if ([_viewControllers count] == 0 && !_allowHeadlessExecution) {
+    [self shutDownEngine];
+  }
+}
+
+- (FlutterViewController*)viewControllerForId:(uint64_t)viewId {
+  FlutterViewController* controller = [_viewControllers objectForKey:@(viewId)];
+  NSAssert(controller == nil || controller.id == viewId,
+           @"The stored controller has unexpected view ID.");
+  return controller;
+}
+
 - (void)setViewController:(FlutterViewController*)controller {
-  if (_viewController == controller) {
+  FlutterViewController* currentController =
+      [_viewControllers objectForKey:@(kFlutterDefaultViewId)];
+  if (currentController == controller) {
     // From nil to nil, or from non-nil to the same controller.
     return;
   }
-  if (_viewController == nil && controller != nil) {
+  if (currentController == nil && controller != nil) {
     // From nil to non-nil.
     NSAssert(controller.engine == nil,
              @"Failed to set view controller to the engine: "
              @"The given FlutterViewController is already attached to an engine %@. "
              @"If you wanted to create an FlutterViewController and set it to an existing engine, "
-             @"you should create it with init(engine:, nibName, bundle:) instead.",
+             @"you should use FlutterViewController#init(engine:, nibName, bundle:) instead.",
              controller.engine);
-    _viewController = controller;
-    [_viewController attachToEngine:self withId:kFlutterDefaultViewId];
-  } else if (_viewController != nil && controller == nil) {
+    [self registerViewController:controller forId:kFlutterDefaultViewId];
+  } else if (currentController != nil && controller == nil) {
     // From non-nil to nil.
-    [_viewController detachFromEngine];
-    _viewController = nil;
-    if (!_allowHeadlessExecution) {
-      [self shutDownEngine];
-    }
+    [self deregisterViewControllerForId:kFlutterDefaultViewId];
```

Comment:
Should this use the `currentController`'s ID?

Reply:
The `setViewController` method should only operate on the default view ID. The `currentController` is bound to have the default view ID, although yeah I can assert here.

Reply:
Good point. An assertion and maybe a quick comment seems like a good idea :)